integration-tests/artemis-jms/src/main/java/io/quarkiverse/messaginghub/pooled/jms/it/CustomConnectionFactory.java

```diff
@@ -5,11 +5,12 @@
 import org.apache.activemq.artemis.jms.client.ActiveMQConnectionFactory;
 import org.eclipse.microprofile.config.ConfigProvider;
 
+import io.quarkiverse.messaginghub.pooled.jms.ComposedConnectionFactory;
 import io.quarkus.arc.properties.UnlessBuildProperty;
 
 @Dependent
 @UnlessBuildProperty(name = "quarkus.artemis.enabled", stringValue = "true")
-public class CustomConnectionFactory extends ActiveMQConnectionFactory {
+public class CustomConnectionFactory extends ActiveMQConnectionFactory implements ComposedConnectionFactory {
```

Comment:
So what happens if `quarkus.artemis.enabled=true` and this bean gets disabled? Does a user have to provide a bean that implements `ConnectionFactory`, or `XAConnectionFactory`, or both?

I'm sorry if that's a stupid question I'm obviously missing a lot of context ;-).

Reply:
No, if `quarkus.artemis.enabled=true`, users don't need to provide any bean and `quarkus-artemis` will produce a `ConnectionFactory` bean which is wrapped with `quarkus-pooled-jms` by using `ConnectionFactoryWrapperBuildItem`.

Reply:
Then it's probably ok.

Reply:
I think that introduction of ComposedConnectionFactory interface will cause some problems. Users will be forced to implement this interface for CF otherwise they will not be wrapped for pooling. Second problem is how to handle CF which do not implement XAConnectionFactory interface. For example [this](https://github.com/apache/camel-quarkus/blob/main/integration-tests/jms-ibmmq-client/src/main/java/org/apache/camel/quarkus/component/jms/ibmmq/it/IBMMQConnectionFactory.java) IBM MQ CF. If users will implement dummy XA methods will it cause problem to transaction manager?

Reply:
Maybe it will be better to remove decorator feature and just wrap automatically only beans via ConnectionFactoryWrapperBuildItem. The cases when we have not got extension for messaging driver like IBM MQ, we create example how to wrap it for pooling. Here is [example](https://github.com/zhfeng/camel-quarkus/compare/issue_5212_ibmmq_xa...osmman:camel-quarkus:issue_5212_ibmmq_xa-tturek?expand=1)

Reply:
Good idea! I agree that decorator feature has much limitations here.

Does your example work for XA transaction?

Reply:
XA tests passed from that example and I see in debugger that methods from JmqiXAResource are called thought pooling's classes.

Reply:
Yeah, that's really a good news.

So all of us agree to remove `Decorator` feature because

- It can only apply to class bean but not produce bean and synethic bean
- It is not possible to wrap `XAConnectionFactory` in case the decorator for `ConnectionFactory` does not expect the decorated bean to also implement `XAConnectionFactory`
- Interception of self-invocation method leads to a recurive loop (even if we can check it by using `StackWalker` API, I think this is still a problem)

And we can just wrap automatically only beans via ConnectionFactoryWrapperBuildItem which should work with `quarkus-artemis` and `quakrus-qpid-jms`. For those messaging drivers like IBM MQ which we have not got extension for, it has to use `PooledJmsWrapper` to wrap the `ConnectionFactory` by themselves.

Is it correctly? or is there anything I missed?

Reply:
@mkouba - Thank you so much for your great helping from the CDI perspective!

@osmman - I'm also thinking about to produce a `PooledJmsWrapper` bean which could be used by custom ConnectionFactory like

```
@Produces
@ApplicationScoped
public ConnectionFactory createConnectionFactory(PooledJmsWrapper wrapper) {
    MQXAConnectionFactory mqCF = new MQXAConnectionFactory();
   mqCF.setPort();
   ...
   return wrapper.wrapConnectionFactory(mqCF);
}
```

`PooledJmsWrapper` should be thread-safe and the `transaction` parameter is used for checking the `quarkus-narayana-jta` extension is involved at build time. WDYT?

Reply:
I open a new PR for reviwing.
- https://github.com/quarkiverse/quarkus-pooled-jms/pull/157

Reply:
Yes, I think it's a good idea